File: dawn_haskell/src/Lsh.hs
module LSHHash
  ( hash
  , hammingDistance
  , lshFunc
  , buildLSHTable
  , queryLSHTable
  ) where

import qualified Data.HashTable.IO as H
import Data.List (nub)
import Data.Char (ord)

hash :: String -> Int
hash = sum . map ord

hammingDistance :: String -> String -> Int
hammingDistance str1 str2 = length $ filter (uncurry (/=)) $ zip str1 str2

lshFunc :: Int -> Int -> String -> [Int]
lshFunc k l str = [hash $ take l $ drop (i * len `div` k) str | i <- [0..k-1]]
  where len = length str

buildLSHTable :: Int -> Int -> [String] -> IO (H.HashTable Int [String])
buildLSHTable k l strings = do
  table <- H.new
  mapM_ (\str -> mapM_ (\hashVal -> H.insert table hashVal [str]) $ lshFunc k l str) strings
  return table

queryLSHTable :: Int -> Int -> H.HashTable Int [String] -> String -> IO [String]
queryLSHTable k l table queryStr = do
  let hashes = lshFunc k l queryStr
  candidates <- concat <$> mapM (H.lookup table) hashes
  return $ filter (\str -> hammingDistance str queryStr <= l) $ nub candidates

File: dawn_python/lsh.py
class LSHHash:
    def hash_str(s):
        return sum(map(ord, s))

    def hamming_distance(s1, s2):
        return sum(c1 != c2 for c1, c2 in zip(s1, s2))

    def lsh_func(k, l, s):
        n = len(s)
        hashes = []
        for i in range(k):
            start = i * n // k
            end = start + l * n // k
            hashes.append(LSHHash.hash_str(s[start:end]))
        return hashes

    def build_lsh_table(k, l, strings):
        table = {}
        for s in strings:
            hashes = LSHHash.lsh_func(k, l, s)
            for h in hashes:
                table.setdefault(h, []).append(s)
        return table

    def query_lsh_table(k, l, table, query_str):
        candidates = []
        hashes = LSHHash.lsh_func(k, l, query_str)
        for h in hashes:
            for s in table.get(h, []):
                if LSHHash.hamming_distance(s, query_str) <= l:
                    candidates.append(s)
        return candidates